boilerplate.py

```diff
@@ -50,15 +50,19 @@
 @_autogen_docstring(Axes.%(func)s)
 def %(func)s(%(argspec)s):
     %(ax)s = gca()
-    # allow callers to override the hold state by passing hold=True|False
-    %(washold)s = %(ax)s.ishold()
+    # Deprecated: allow callers to override the hold state
+    # by passing hold=True|False
+    %(washold)s = %(ax)s._hold
 %(sethold)s
     if hold is not None:
-        %(ax)s.hold(hold)
+        %(ax)s._hold = hold
+        from matplotlib.cbook import mplDeprecation
+        warnings.warn("The 'hold' keyword argument is deprecated since 2.0.",
+                      mplDeprecation)
```

Comment:
This is with ipython and stacklevel=2:
```
In [1]: plot([1,2,3], hold=True)
/Users/efiring/anaconda/envs/test/bin/ipython:1: MatplotlibDeprecationWarning: The 'hold' keyword argument is deprecated since 2.0.
  #!/bin/bash /Users/efiring/anaconda/envs/test/bin/python.app
[<matplotlib.lines.Line2D at 0x1151f1f98>]

In [2]: def make_warning():
   ...:     plot([1,2,3], hold=True)
   ...:

In [3]: make_warning()
/Users/efiring/anaconda/envs/test/bin/ipython:2: MatplotlibDeprecationWarning: The 'hold' keyword argument is deprecated since 2.0.
  if __name__ == '__main__':
```
and with the default:
```
In [1]: plot([1,2,3], hold=True)
/Users/efiring/work/programs/py/mpl/matplotlib/lib/matplotlib/pyplot.py:3315: MatplotlibDeprecationWarning: The 'hold' keyword argument is deprecated since 2.0.
  mplDeprecation)
[<matplotlib.lines.Line2D at 0x117dce0b8>]

In [2]: def make_warning():
    plot([1,2,3], hold=True)
   ...:

In [3]: make_warning()
/Users/efiring/work/programs/py/mpl/matplotlib/lib/matplotlib/pyplot.py:3315: MatplotlibDeprecationWarning: The 'hold' keyword argument is deprecated since 2.0.
  mplDeprecation)
```
For this use, the default looks better to me.  What would really be needed for someone to debug a hold kwarg buried in a library is a full stack trace, but we are not going to put that in.